cpp/src/gandiva/dex.h

```diff
@@ -272,6 +275,73 @@ class BooleanOrDex : public BooleanDex {
   void Accept(DexVisitor& visitor) override { visitor.Visit(*this); }
 };
 
+// decomposed in expression.
+template <typename Type>
+class InExprDex;
+
+template <typename Type>
+class InExprDexBase : public Dex {
+ public:
+  explicit InExprDexBase(const ValueValidityPairVector& args) : args_(args) {}
+
+  const ValueValidityPairVector& args() const { return args_; }
+
+  void Accept(DexVisitor& visitor) override { visitor.Visit(*this); }
+
+  const std::string& runtime_function() const { return runtime_function_; }
+
+ protected:
+  ValueValidityPairVector args_;
+  std::string runtime_function_;
+};
+
+template <>
+class InExprDex<int32_t> : public InExprDexBase<int32_t> {
```

Comment:
The three specializations look almost the same (except the name of the runtime function). Can't you use common code for them?

Reply:
yes i should! moved everything other than runtime function to base class..